cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -227,23 +229,39 @@ private predicate storeStepChi(Node node1, Content f, PostUpdateNode node2) {
       c = chi.getResultType() and
       exists(int startBit, int endBit |
         chi.getUpdatedInterval(startBit, endBit) and
-        f.(FieldContent).hasOffset(c, startBit, endBit)
+        f.hasOffset(c, startBit, endBit)
       )
       or
-      getWrittenField(store, f.(FieldContent).getAField(), c) and
-      f.(FieldContent).hasOffset(c, _, _)
+      getWrittenField(store, f.getAField(), c) and
+      f.hasOffset(c, _, _)
     )
   )
 }
 
+private predicate arrayStoreStepChi(Node node1, ArrayContent a, PostUpdateNode node2) {
+  a = TArrayContent() and
+  exists(StoreInstruction store |
+    node1.asInstruction() = store and
+    (
+      // `x[i] = taint()`
+      store.getDestinationAddress() instanceof PointerAddInstruction
+      or
+      // `*p = taint()`
+      store.getDestinationAddress().(CopyValueInstruction).getUnary() instanceof LoadInstruction
```

Comment:
These rules seem overly syntactic, but we can always tweak them later if there's a need. The important thing is they're guaranteed not to overlap with `getWrittenField`.

Reply:
I agree. The `PointerAdd` case I think we can live with, but requiring the presence of a `CopyValue` instruction as a direct successor to the `Load` instruction is quite fragile. For instance, for this store:
```
int source();
void f(char* s) {
	*(int*)s = source();
}
```
there's a `Convert` instruction in between the `Load` and the `CopyValue`, which means that this is not recognized as a store step by `arrayStoreStepChi`.